File: apps/sample-api/src/intents/promoted.ts
import { formatQueryRef } from "@kohaku-ui/data-binding";
import { compileQueryTemplate } from "@kohaku-ui/intents";
import type { ComponentDraft } from "@kohaku-ui/lineage";
import { type ComponentDefinition, defineComponent, propsSchemaFromJsonSchema } from "@kohaku-ui/registry";
import { z } from "zod";
import type { IntentDef } from "./catalog.js";
import { fiscalYear, region } from "./vocab.js";

/**
 * The artifact of promotion (L2->L1 publish) (an element of the projection held by the per-tenant PromotedRegistry).
 * - Component: registered in the catalog as a sandbox-template implementation (promotion = coming under governance;
 *   replacement with a native implementation is a human development task, and the Web side renders natively once the
 *   implementation is registered)
 * - Intent: joins the natural-language-normalization vocabulary (merged into the LLM prompt catalog)
 * - Persistence: the promotion-state snapshot (promotions.json) is the sole state authority, and this projection
 *   is rebuilt by startup reconcile from the published snapshot + component.generated (html).
 *   Therefore no independent persistence to promoted.json is done (the snapshot is authoritative).
 */
export interface PromotedEntry {
  artifactId: string;
  draft: ComponentDraft;
  html: string;
  request?: string;
  publishedAt: string;
}

export function promotedComponent(entry: PromotedEntry): ComponentDefinition {
  // If draft.paramsJsonSchema exists, reconstruct propsSchema from it. As a total conversion it does not
  // throw at startup. defineComponent requires propsSchema to be a ZodObject, but propsSchemaFromJsonSchema may
  // return a z.record for a non-object schema. At runtime, safeParse / toJSONSchema work with either (the
  // sandbox-template does not use ZodObject-specific operations of L1 generation), so we treat it as a ZodObject
  // only at the boundary.
  const propsSchema = (
    entry.draft.paramsJsonSchema != null
      ? propsSchemaFromJsonSchema(entry.draft.paramsJsonSchema)
      : z.object({ title: z.string().optional() })
  ) as z.ZodObject;
  return defineComponent({
    type: entry.draft.componentType,
    version: entry.draft.version,
    description: entry.draft.description,
    propsSchema,
    capabilities: { events: [], data: "required", children: "none" },
    implementation: { kind: "sandbox-template", html: entry.html },
    fallback: {
      type: "presentMarkdown",
      mapProps: () => ({ markdown: `(${entry.draft.componentType} is not available on this surface)` }),
    },
  });
}

/**
 * Default Intent params for old promoted.json compatibility (no paramsJsonSchema / queryTemplate).
 * The value ranges for fiscalYear / region reference the same source as catalog (vocab.ts) (eliminating the duplicate definition).
 */
const DEFAULT_PROMOTED_PARAMS = z.object({
  fiscalYear: fiscalYear.default(2026),
  region: region.enum().optional(),
});

export function promotedIntent(entry: PromotedEntry): IntentDef {
  const { draft } = entry;
  // GUI facet input arrives as strings, so number/boolean are received via the coerce variant (paired with the
  // propsSchema validation, intent params use input-tolerant coercion). If paramsJsonSchema is not an object,
  // fall back to the default params (fiscalYear/region).
  const params = coercedParamsSchema(draft.paramsJsonSchema) ?? DEFAULT_PROMOTED_PARAMS;
  const template = draft.queryTemplate;
  return {
    name: draft.intentName,
    description: draft.description,
    params,
    examples: [entry.request ?? draft.description],
    toQueries: (p) => {
      if (template != null) {
        // Map intent params -> query params along the queryTemplate.
        // Unified on the same @kohaku-ui/intents helper as core Intents (a single implementation of canonicalization and missing-value exclusion).
        return [compileQueryTemplate("sales", template, p)];
      }
      // Old promoted.json compatibility: fixed trend logic.
      return [
        {
          uri: formatQueryRef({
            source: "sales",
            path: "trend",
            params: {
              fy: String(p["fiscalYear"] as number),
              ...(p["region"] != null ? { region: String(p["region"]) } : {}),
              metric: "revenue",
              granularity: "month",
            },
          }),
        },
      ];
    },
  };
}

/**
 * Builds a coerce schema for intent params from paramsJsonSchema (object).
 * Similar to propsSchemaFromJsonSchema, but to allow GUI-originated string input it uses z.coerce.number for
 * number/integer and z.coerce.boolean for boolean (query params are assumed to be flat primitives only).
 * Returns null for non-object / missing (the caller falls back to the default params).
 */
function coercedParamsSchema(schema: unknown): z.ZodObject | null {
  if (
    schema == null ||
    typeof schema !== "object" ||
    Array.isArray(schema) ||
    (schema as Record<string, unknown>)["type"] !== "object"
  ) {
    return null;
  }
  const properties = (schema as Record<string, unknown>)["properties"];
  if (properties == null || typeof properties !== "object") return null;
  const requiredList = Array.isArray((schema as Record<string, unknown>)["required"])
    ? ((schema as Record<string, unknown>)["required"] as unknown[])
    : [];
  const required = new Set(requiredList.filter((k): k is string => typeof k === "string"));
  const shape: Record<string, z.ZodType> = {};
  for (const [key, node] of Object.entries(properties as Record<string, unknown>)) {
    const base = coerceBase(node);
    const n = node as Record<string, unknown>;
    if (n != null && typeof n === "object" && "default" in n) {
      shape[key] = base.default(n["default"] as never);
    } else {
      shape[key] = required.has(key) ? base : base.optional();
    }
  }
  return z.object(shape);
}

/** Receives flat primitives via the coerce variant (unsupported types become z.unknown). */
function coerceBase(node: unknown): z.ZodType {
  if (node == null || typeof node !== "object") return z.unknown();
  const n = node as Record<string, unknown>;
  switch (n["type"]) {
    case "string": {
      const en = n["enum"];
      if (Array.isArray(en) && en.length > 0 && en.every((x) => typeof x === "string")) {
        return z.enum(en as [string, ...string[]]);
      }
      return z.string();
    }
    case "number":
      return z.coerce.number();
    case "integer":
      return z.coerce.number().int();
    case "boolean":
      // Allow GUI-originated string input. Because z.coerce.boolean has Boolean(input) semantics,
      // "false"/"0" would turn into true as non-empty strings. Receive via an explicit mapping:
      // "false"/"0"/false -> false, "true"/"1"/true -> true, everything else is rejected by z.boolean() validation.
      return z.preprocess((v) => {
        if (v === "true" || v === "1" || v === true) return true;
        if (v === "false" || v === "0" || v === false) return false;
        return v;
      }, z.boolean());
    default:
      return z.unknown();
  }
}

File: python/examples/sales-api/src/sales_api/promoted.py
"""Mapping of promotion (L2->L1 publish) artifacts -> component / Intent (port of TS: apps/sample-api/src/intents/promoted.ts).

Promotion = entering governance. A promoted component is registered in the catalog as a sandbox-template implementation,
and a promoted Intent is added to the vocabulary of natural-language normalization.

This module holds pure transforms (promoted_component / promoted_intent) and coerce-schema construction. The per-tenant
registry (holding / rebuilding the projection) is promoted_registry.py; reconcile and the lineage / snapshot integration
are handled by app.py's wiring layer (create_promotions' on_publish / reconcile) (symmetric with TS: promoted-registry.ts / app.ts).

The param coerce uses the same mapping as TS (apps/sample-api/src/intents/promoted.ts's coerceBase): number/integer are
coerce number, boolean is an explicit mapping (equivalent to z.coerce.boolean), and other types or non-object nodes pass
through (equivalent to z.unknown()). boolean / unknown are not in the base DSL, so they are supplemented within this
module by subclassing ParamField (kohaku.intents remains unchanged).
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any

from kohaku.data_binding import format_query_ref
from kohaku.intents import (
    EnumField,
    IntentDef,
    ObjectSchema,
    ParamError,
    ParamField,
    compile_query_template,
    number,
    object_schema,
    string,
)
from kohaku.intents import QueryTemplate as IntentQueryTemplate
from kohaku.lineage import ComponentDraft
from kohaku.registry import (
    CapabilityDecl,
    ComponentDefinition,
    FallbackDecl,
    ImplementationDecl,
    PropsSchema,
    define_component,
    props_schema_from_json_schema,
)
from kohaku.spec import JsonObject, QueryHandle, js_string

from .domain import DEMO_FISCAL_YEAR
from .intents_catalog import fiscal_year, region


@dataclass(frozen=True)
class PromotedEntry:
    """A promotion entry (an element of the projection held by the per-tenant registry).

    Persistence: the promotion-state snapshot (promotions.json) is the sole state authority, and this projection is
    rebuilt by startup reconcile (no independent persistence to promoted.json). reconcile is the responsibility of the
    host wiring layer.
    """

    artifactId: str
    draft: ComponentDraft
    html: str
    request: str | None = None
    publishedAt: str = ""


def promoted_component(entry: PromotedEntry) -> ComponentDefinition:
    """Turns a promoted component into a ComponentDefinition for catalog registration as a sandbox-template implementation."""
    # If draft.paramsJsonSchema exists, reconstruct propsSchema from it (a total transform, so it does not throw at startup).
    props_schema = (
        props_schema_from_json_schema(entry.draft.paramsJsonSchema)
        if entry.draft.paramsJsonSchema is not None
        else PropsSchema({"type": "object", "properties": {"title": {"type": "string"}}})
    )
    component_type = entry.draft.componentType

    def _fallback(_props: JsonObject) -> JsonObject:
        return {"markdown": f"({component_type} is not available on this surface)"}

    return define_component(
        ComponentDefinition(
            type=component_type,
            version=entry.draft.version,
            description=entry.draft.description,
            propsSchema=props_schema,
            capabilities=CapabilityDecl(events=[], data="required", children="none"),
            implementation=ImplementationDecl(kind="sandbox-template", html=entry.html),
            fallback=FallbackDecl(type="presentMarkdown", map_props=_fallback),
        )
    )


# Default Intent params for old promoted.json compatibility (without paramsJsonSchema / queryTemplate).
# The value ranges of fiscalYear / region reference the same source as catalog (intents_catalog).
DEFAULT_PROMOTED_PARAMS: ObjectSchema = object_schema(
    {"fiscalYear": fiscal_year.default(DEMO_FISCAL_YEAR), "region": region.enum().optional()}
)


def promoted_intent(entry: PromotedEntry) -> IntentDef:
    """Turns a promoted Intent into an IntentDef for SemanticPort."""
    draft = entry.draft
    # GUI facet input arrives as strings, so it is received with the coerce version. If paramsJsonSchema is not an object, the default params.
    params = _coerced_params_schema(draft.paramsJsonSchema) or DEFAULT_PROMOTED_PARAMS
    template = draft.queryTemplate

    def to_queries(p: JsonObject) -> list[QueryHandle]:
        if template is not None:
            # Unified with the same compile_query_template as core Intents (a single implementation of canonicalization / missing-value exclusion).
            # lineage's QueryTemplate is structurally identical to intents' QueryTemplate but a different type, so it is converted.
            intent_template = IntentQueryTemplate(
                path=template.path, paramMap=template.paramMap, fixedParams=template.fixedParams
            )
            return [compile_query_template("sales", intent_template, p)]
        # Old promoted.json compatibility: fixed trend logic.
        query_params = {"fy": js_string(p["fiscalYear"])}
        if p.get("region") is not None:
            query_params["region"] = js_string(p["region"])
        query_params["metric"] = "revenue"
        query_params["granularity"] = "month"
        return [QueryHandle(uri=format_query_ref(source="sales", path="trend", params=query_params))]

    return IntentDef(
        name=draft.intentName,
        description=draft.description,
        params=params,
        examples=[entry.request if entry.request is not None else draft.description],
        to_queries=to_queries,
    )


def _coerced_params_schema(schema: Any) -> ObjectSchema | None:
    """Builds a coerce schema for intent params from paramsJsonSchema (object).

    To allow GUI-origin string input, number/integer are received as coerce number. None if non-object or missing.
    """
    if not isinstance(schema, dict) or schema.get("type") != "object":
        return None
    properties = schema.get("properties")
    if not isinstance(properties, dict):
        return None
    required_list = schema.get("required")
    required = (
        {k for k in required_list if isinstance(k, str)}
        if isinstance(required_list, list)
        else set()
    )
    shape: dict[str, ParamField] = {}
    for key, node in properties.items():
        base = _coerce_base(node)
        if isinstance(node, dict) and "default" in node:
            shape[key] = base.default(node["default"])
        else:
            shape[key] = base if key in required else base.optional()
    return object_schema(shape)


@dataclass(frozen=True)
class _BooleanField(ParamField):
    """Equivalent to z.coerce.boolean. Receives GUI-origin string input via an explicit mapping (query params are strings).

    Because z.coerce.boolean (= Boolean(input) semantics) turns "false"/"0" into true as non-empty strings, the TS side
    (promoted.ts) does an explicit mapping via z.preprocess. Matching that:
    a boolean passes through / "true"・"1" -> True / "false"・"0" -> False / anything else is a validation error.
    """

    def _coerce(self, value: Any) -> Any:
        if isinstance(value, bool):
            return value
        if value in ("true", "1"):
            return True
        if value in ("false", "0"):
            return False
        raise ParamError(f"not a boolean: {value!r}")


@dataclass(frozen=True)
class _UnknownField(ParamField):
    """Equivalent to z.unknown(). Passes through without coercing or validating (a catch-all for unsupported types / non-object nodes)."""

    def _coerce(self, value: Any) -> Any:
        return value


def _coerce_base(node: Any) -> ParamField:
    """Receives a flat primitive with a coerce-version field (the same mapping as TS's coerceBase)."""
    if not isinstance(node, dict):
        return _UnknownField()  # TS: if node is non-object, z.unknown()
    node_type = node.get("type")
    if node_type == "string":
        enum_values = node.get("enum")
        if (
            isinstance(enum_values, list)
            and len(enum_values) > 0
            and all(isinstance(x, str) for x in enum_values)
        ):
            return EnumField(values=tuple(enum_values))
        return string()
    if node_type == "number":
        return number()
    if node_type == "integer":
        return number(integer=True)
    if node_type == "boolean":
        return _BooleanField()
    # Unsupported types are equivalent to z.unknown() (pass through without coercing or validating).
    return _UnknownField()


__all__ = [
    "DEFAULT_PROMOTED_PARAMS",
    "PromotedEntry",
    "promoted_component",
    "promoted_intent",
]
